### DISCUSSION TITLE
Help with adding custom python library to a course

### DISCUSSION THREAD

**Discussion by justindr21** [PrairieLearn/PrairieLearn#6044] [Q&A] [answered]
Hi, I am trying to import a custom library to be used within a course. Upon adding the package into serverFilesCourse, I got this error:

`ImportError: cannot import name '_psutil_linux' from partially initialized module 'psutil' (most likely due to a circular import) (/grade/serverFilesCourse/psutil/__init__.py)`

I have been advised to create a custom image for the course instead as per the replies from these two wonderful people on the PrairieLearn Slack: 
https://prairielearn.slack.com/archives/C266KEH9A/p1657793474396089
https://prairielearn.slack.com/archives/C266KEH9A/p1659019377617369

And as of now I am still working through the instructions from the first link but I wanted to create this post anyways just for the sake of it and getting comfortable with posting here instead of Slack since it is a little scarier than in Slack. I also expect to run into more problems which is the real reason I have decided to make this discussion. 

Note: It has been decided that instead of using a file-editor, a jupyter notebook workspace would be much better to use for the course that I am trying to enable the library for and just for context, it would be a great start to just be able to let the student use the library within a jupyter notebook workspace in a question and run their code.

**Comment by ffund** [comment, marked as answer]
In general, there are a couple of ways to use an additional library in the Python auto-grader:

1. install the library in the course's `serverFilesCourse` directory. Then add a `serverFilesCourse` option in the question's `externalGradingOptions` as described [here](https://prairielearn.readthedocs.io/en/latest/externalGrading/#configuring-and-enabling-external-grader-support).
2. make your own external grader image based on the `prairielearn/grader-python` image, but with additional libraries installed, and then point to your image in the `externalGradingOptions`.

In my experience, 

* option 1 (putting stuff in `serverFilesCourse`  ) works well for pure Python stuff especially if there are not many dependencies, 
* option 2 (make your own grader image) works better for packages with many dependencies or with shared library files. (but the disadvantage is that then you have to maintain your own version of the image, which is a pain.)

it looks like `psutil` has those .so files so you might be better off with option 2.

Here's my quick guide to option 2 (making your own grader image):

* download docker, and sign up on docker hub if you don't already have an account. create a public repo on docker hub. in this example, my account is `ffund` and I have a public repo called `grader-python`.
* clone a copy of the PL github repository. navigate to `graders/python`  - that is the PL Python grader's base image, and you can build yours off of that. 
* edit` requirements.txt` if you need to add additional packages that you want to have available to the grader.
* then, in that directory,
```
docker build -t ffund/grader-python .
# probably a good idea to run your container to test at this stage
docker login
docker push ffund/grader-python
```
* now in your question, you can set the image in the `externalGradingOptions` to your image e.g. `ffund/grader-python`
* note that in the course "sync" page you can pull the latest version of your image from docker hub

(note that I used my own username and image name in that example, replace with your own)

**Reply by ffund** [reply]
That's for the grader - if you need the library to also be available in the Jupyter workspace, then you'll do something similar but in `workspaces/jupyterlab` in the PL base code.

An alternative option for workspaces, if I don't feel like maintaining a separate workspace image, is to put a `whl` version of the library in the workspace (so that it can be installed without networking), install it in the first cell of the notebook, then direct students to restart the kernel. e.g.

![image](https://user-images.githubusercontent.com/2316553/181802731-81337e00-f8ed-4425-a68c-72e8a61c1194.png)

**Reply by jonatanschroeder** [reply]
Another option for custom Docker containers which may make it simpler to maintain is to create a Dockerfile that is based on PL's version itself, something like:

```
FROM prairielearn/grader-python
RUN pip3 install <list your packages here>
```

and follow that with @ffund's build commands from above. I've done this in the past (with the Java autograder, though). In my personal view this has the advantage that, to update your custom image based on updates in the PL autograder, you don't need to keep close track of changed files in the PL version, you only need to build/push and it will synchronize with the newer changes.

**Reply by coatless** [reply]
For a slightly more automated version of @ffund's option 2, c.f.

https://github.com/stat-prairielearn-uiuc/grader-dl
https://github.com/stat-prairielearn-uiuc/workspace-dl-jupyterlab

**Reply by justindr21** [reply]
Is it possible to run `import sys` and `!{sys.executable} -m pip install <package>` once the workspace loads without the student seeing anything?

**Reply by jonatanschroeder** [reply]
I'm personally not sure if it's possible (I don't have a lot of experience with the Jupyter workspace), but that would require you to enable networking in the workspace, which is typically a bad idea (see https://prairielearn.readthedocs.io/en/latest/workspaces/#infojson). So if you can do it in your custom workspace is probably better.

**Reply by echuber2** [reply]
I know of one course that did a hacky post-install like that to get around an issue, but it's not really advisable. If you do make a custom workspace based off of the PL base image(s), you'll want to install any additional packages in the image itself (baked in), not in the entrypoint script. (Doing a lot of heavy lifting during the entrypoint would _still_ require network access, and it makes the workspace load slowly, or even time out.)

**Reply by justindr21** [reply]
Thanks for the input! I tried looking into making the custom workspace image and was wondering if there is an example of how to set entrypoints as well as "baking in packages" in the image?

**Reply by echuber2** [reply]
This assumes some familiarity with Docker. What you would want to do is start with something like the existing JupyterLab workspace, then build and push to your own Docker Hub account so that your course can sync from your own image: https://github.com/PrairieLearn/PrairieLearn/tree/master/workspaces/jupyterlab

The simplest thing would be to edit the contents of `install.sh` and rebuild the Docker image, but take note that this installs packages as the root user. Also, I see a few issues with the existing Dockerfile we have (it's a bit out of date and the user `jovyan` maps to 1000, despite us running the image as 1001).

You must be sure that the files are installed in the image itself, not by some runtime script invoked by the entrypoint to the container (as specified by `ENTRYPOINT` or `CMD`).

**Comment by justindr21** [comment]
Thank you both so much! I was able to create the custom image, however I still getting that exact same error I mentioned above. I am also unsure what the course "sync" page may be referring to. I also want to know that if it says `Pulling image justindr21/grader-python` on my terminal where I put the docker run command, that I most likely have done the right things to make the custom image?

**Reply by ffund** [reply]
@justindr21 this is the "Sync" page and this is how you sync your custom image:

![image](https://user-images.githubusercontent.com/2316553/182910276-99afab90-38b3-4989-b345-daeedec15efb.png)

**Reply by jonatanschroeder** [reply]
> however I still getting that exact same error I mentioned above

I'm guessing that if you have the Python package on both serverFilesCourse and the docker image then they may be interpreted in an order that is different than what you expect. If you still have it in serverFilesCourse, then I'd suggest you remove it.

>  I am also unsure what the course "sync" page may be referring to.

Once you link a question to this new repo in the question's info.json in the official PrairieLearn page (not your local installation), then the Sync page of PrairieLearn will keep a list of Docker images used in your questions, including the new image. This list is cached, so before you use it for the same time or if you change it, you need to synchronize it in the Sync page.

<img width="1253" alt="image" src="https://user-images.githubusercontent.com/1004907/182910810-c694e97d-9c3c-4ce8-ad4a-a4b26744abe4.png">

> I also want to know that if it says Pulling image justindr21/grader-python on my terminal where I put the docker run command, that I most likely have done the right things to make the custom image?

It means that you have properly assigned the question to the correct question, and assuming no errors show up that the image is properly in Docker Hub.

**Reply by justindr21** [reply]
Ah, removing it from serverFilesCourse allows me to import the library now, thanks!

**Reply by justindr21** [reply]
So just to make sure I understand, anything to do with syncing should be done when things are ready to be put on the public, official prairielearn website is that correct? Since I currently don't have authority to tinker with things on the official prairielearn website, I'm assuming that once I merge the changes (which should only be the change of the `image` element in `serverFilesCourse` to be my own custom image), then it will show up as an image that I should be able to sync, does that fit with what is actually going on?

**Reply by jonatanschroeder** [reply]
That is roughly the idea, yes.
